### ISSUE TITLE
HP3 Reversible Color Transform not lossless

### ISSUE THREAD

**Issue by guybonneau** [team-charls/charls#48]
When using the HP3 color transform the encoding to decoding with lossless doesn't yield the same file bit to bit.  Reproducing steps:

1- Use CharLSTest
2- Add this parameter to JlsParameters to function EncodePnm params.colorTransformation = charls::ColorTransformation::HP3;
3- encodepnm with the attached ppm file
4- decodetopnm the compressed file
5- comparepnm fails.

[Land10.zip](https://github.com/team-charls/charls/files/3086287/Land10.zip)

**Comment by vbaderks** [comment]
Hi Guy,

Thanks for the defect report, will take a look at it. Rather busy at the moment, so it can take some time. Will keep you updated.

**Comment by vbaderks** [comment]
Note: I am able to reproduce the problem, No color transformation and ColorTransformation::HP1 works fine, but ColorTransformation::HP2 and ColorTransformation::HP3 seems to be broken.

**Comment by vbaderks** [comment]
I have analysed the problem. The used sample image is a 10 bit RGB image, the color transformation are however only designed for 8 bit and 16 bit RGB images. There is a mechanism in place to bit shift non 8.16 bit images. This shift however is however not the algorithm. 
I investigated the original HP codec , it also only supports color transformations for 8 bit RGB images. As these color transformations are a HP extension, it doesn't make sense for the interoperability to support other options then 8 bit.
The primary motivation for adding HP color transformations in the past was to decode old JPEG-LS images that were encoded with the HP codec. The plan forward is add a check that color transformation are only supported for 8 bit RGB images.
